### PR TITLE
Added mnist denoising autoencoder example

### PR DESCRIPTION
**Pull request by harshhemani**
Using a custom denoising autoencoder layer.

### CODE REVIEW

**Comment by lukedeo** [comment]
Not to be harsh, but the whole pretraining thing is a bit 2008...there really is no need unless you are working with lots of unlabeled data.

**Comment by harshhemani** [comment]
@lukedeo I was trying to do landcover identification on google maps images, I couldn't find a big enough labelled corpus for that. I tried finetuning the caffe (imagenet pretrained) models but that didn't work with my small labelled dataset (~200 images per class). So I started looking for a theano based solution since I was familiar with it. I couldn't find any pretraining based examples which deal with tied weights in a way I wanted them to be. So I wrote my code and thought others might benefit from an example. Do you think I should've approached the problem differently? I tried using CNN but got stuck at 80% accuracy and couldn't seem to be getting any better from there. Should I focus on building a larger labelled dataset instead of exploring pretraining based models? I am still running tests on my autoencoder so couldn't quote accuracy for it.

Best Regards

**Comment by Ori226** [comment]
@lukedeo 
    "Not to be harsh, but the whole pretraining thing is a bit 2008"
I'm pretty new to the Deep Learning field, can you please explain why do you think that there is no need of pre training? The batch normalization techniques? The different initalization methods "Glorut" for example)? Thanks

**Comment by fchollet** [comment]
@harshhemani so what are you need results with pre-training?

In general, pre-training via autoencoders is not useful, and hasn't been in fashion in years (it was only a handful of people that were into it to begin with...). What people usually mean by "pre-training" these days, is to use a network trained on a different but similar (and larger) dataset and fine-tune it (re-train it with a slower learning rate) on the target dataset. This is a valid approach when you don't have a lot of data.

> ~200 images per class

You are going to need a *very* small convolutional network.

**Comment by harshhemani** [comment]
@fchollet Since fine-tuning existing models didn't work for me, I thought writing code will help me and anyone who faces similar problems in future. Thank you for your insight.
